File: v8-core/src/telemetry/identity.rs
#![allow(dead_code)]
//! Canonical Economic Trace Identity & Provenance (EEO-001H, D-136).
//!
//! Constitutional Invariants:
//! 1. Opportunity Sovereignty: `OpportunityId` identifies the market economic event.
//! 2. Trajectory Identity: `EconomicTraceId` identifies a specific execution or decision path.
//! 3. Provenance Isolation: `TraceProvenance` captures the cryptographic commit/tape/policy state
//!    without destroying identity alignment between baseline, challenger, and counterfactual runs.
//! 4. Typed Modality: `TrajectoryType` distinguishes `Observed` from `Counterfactual` without naming heuristics.

use std::fmt;
use std::ops::Deref;
use serde::{Deserialize, Serialize};
use crate::error::V8CoreError;
use crate::hash::Canon;
use crate::opportunity::book::OpportunityEpisode;

/// Execution modality for an economic decision trajectory.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub enum TrajectoryType {
    /// Canonical observed forward simulation or physical execution.
    Observed,
    /// Counterfactual branch under registered or exploratory policy intervention.
    Counterfactual,
}

impl TrajectoryType {
    pub fn is_observed(&self) -> bool {
        matches!(self, Self::Observed)
    }

    pub fn is_counterfactual(&self) -> bool {
        matches!(self, Self::Counterfactual)
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Observed => "Observed",
            Self::Counterfactual => "Counterfactual",
        }
    }
}

/// Cryptographic provenance context binding the decision environment (D-136-RP-001 §6.1).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct TraceProvenance {
    pub tape_hash: String,
    pub policy_hash: String,
    pub constitution_hash: String,
    pub code_hash: String,
}

impl TraceProvenance {
    pub fn new(
        tape_hash: impl Into<String>,
        policy_hash: impl Into<String>,
        constitution_hash: impl Into<String>,
        code_hash: impl Into<String>,
    ) -> Result<Self, V8CoreError> {
        let tape_hash = tape_hash.into();
        let policy_hash = policy_hash.into();
        let constitution_hash = constitution_hash.into();
        let code_hash = code_hash.into();

        if tape_hash.is_empty() {
            return Err(V8CoreError::TraceLineageError(
                "tape_hash cannot be empty in TraceProvenance".to_string(),
            ));
        }
        if policy_hash.is_empty() {
            return Err(V8CoreError::TraceLineageError(
                "policy_hash cannot be empty in TraceProvenance".to_string(),
            ));
        }
        if constitution_hash.is_empty() {
            return Err(V8CoreError::TraceLineageError(
                "constitution_hash cannot be empty in TraceProvenance".to_string(),
            ));
        }
        if code_hash.is_empty() {
            return Err(V8CoreError::TraceLineageError(
                "code_hash cannot be empty in TraceProvenance".to_string(),
            ));
        }

        Ok(Self {
            tape_hash,
            policy_hash,
            constitution_hash,
            code_hash,
        })
    }

    /// Computes deterministic BLAKE3 digest of the provenance bundle.
    pub fn compute_hash(&self) -> String {
        let mut c = Canon::new();
        c.push_str("TraceProvenance");
        c.push_str(&self.tape_hash);
        c.push_str(&self.policy_hash);
        c.push_str(&self.constitution_hash);
        c.push_str(&self.code_hash);
        c.finish_blake3_hex()
    }
}

/// Canonical Economic Trace Identifier (BLAKE3-derived hex).
/// Identifies a specific execution/decision trajectory for an opportunity.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct EconomicTraceId(pub String);

impl EconomicTraceId {
    pub fn new(id: impl Into<String>) -> Self {
        Self(id.into())
    }

    /// Computes deterministic BLAKE3 identity for an economic trajectory.
    /// Trajectory tag distinguishes baseline vs challenger vs counterfactual branches
    /// while preserving opportunity identity alignment.
    pub fn compute(
        opportunity_id: &str,
        trajectory_tag: &str,
        trajectory_type: TrajectoryType,
        pit_timestamp: i64,
    ) -> Self {
        let mut c = Canon::new();
        c.push_str("EconomicTraceId");
        c.push_str(opportunity_id);
        c.push_str(trajectory_tag);
        c.push_str(trajectory_type.as_str());
        c.push_i64(pit_timestamp);
        Self(c.finish_blake3_hex())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for EconomicTraceId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Deref for EconomicTraceId {
    type Target = str;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl AsRef<str> for EconomicTraceId {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<String> for EconomicTraceId {
    fn from(s: String) -> Self {
        Self(s)
    }
}

impl From<&str> for EconomicTraceId {
    fn from(s: &str) -> Self {
        Self(s.to_string())
    }
}

/// Canonical Decision Span Identifier (BLAKE3-derived hex).
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct SpanId(pub String);

impl SpanId {
    pub fn new(id: impl Into<String>) -> Self {
        Self(id.into())
    }

    /// Computes deterministic BLAKE3 identity for a span.
    pub fn compute(
        trace_id: &EconomicTraceId,
        parent_span_id: Option<&SpanId>,
        stage_name: &str,
        start_time: i64,
        disambiguator: &str,
    ) -> Self {
        let mut c = Canon::new();
        c.push_str("SpanId");
        c.push_str(trace_id.as_str());
        if let Some(parent) = parent_span_id {
            c.push_str(parent.as_str());
        } else {
            c.push_null();
        }
        c.push_str(stage_name);
        c.push_i64(start_time);
        c.push_str(disambiguator);
        Self(c.finish_blake3_hex())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for SpanId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Deref for SpanId {
    type Target = str;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl AsRef<str> for SpanId {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<String> for SpanId {
    fn from(s: String) -> Self {
        Self(s)
    }
}

impl From<&str> for SpanId {
    fn from(s: &str) -> Self {
        Self(s.to_string())
    }
}

/// Immutable Root Context of an Economic Trace (D-136-RP-001 §6.1).
/// Preserves opportunity identity, trajectory identity, and provenance as distinct dimensions.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EconomicTraceContext {
    pub trace_id: EconomicTraceId,
    pub opportunity_id: String,
    pub trajectory_type: TrajectoryType,
    pub trajectory_tag: String,
    pub pit_timestamp: i64,
    pub provenance: TraceProvenance,
}

impl EconomicTraceContext {
    /// Creates and cryptographically validates a new `EconomicTraceContext`.
    pub fn new(
        opportunity_id: impl Into<String>,
        trajectory_type: TrajectoryType,
        trajectory_tag: impl Into<String>,
        pit_timestamp: i64,
        provenance: TraceProvenance,
    ) -> Result<Self, V8CoreError> {
        let opportunity_id = opportunity_id.into();
        let trajectory_tag = trajectory_tag.into();

        if opportunity_id.is_empty() {
            return Err(V8CoreError::TraceLineageError(
                "opportunity_id cannot be empty".to_string(),
            ));
        }
        if trajectory_tag.is_empty() {
            return Err(V8CoreError::TraceLineageError(
                "trajectory_tag cannot be empty".to_string(),
            ));
        }

        let trace_id = EconomicTraceId::compute(
            &opportunity_id,
            &trajectory_tag,
            trajectory_type,
            pit_timestamp,
        );

        Ok(Self {
            trace_id,
            opportunity_id,
            trajectory_type,
            trajectory_tag,
            pit_timestamp,
            provenance,
        })
    }

    /// Constructs an `EconomicTraceContext` for an observed opportunity episode.
    pub fn from_episode(
        episode: &OpportunityEpisode,
        tape_hash: &str,
        policy_hash: &str,
        constitution_hash: &str,
        code_hash: &str,
    ) -> Result<Self, V8CoreError> {
        let prov = TraceProvenance::new(tape_hash, policy_hash, constitution_hash, code_hash)?;
        Self::new(
            &episode.episode_id,
            TrajectoryType::Observed,
            "canonical_observed",
            episode.as_of_time,
            prov,
        )
    }

    /// Constructs an `EconomicTraceContext` for a named challenger or counterfactual trajectory.
    pub fn from_episode_trajectory(
        episode: &OpportunityEpisode,
        trajectory_type: TrajectoryType,
        trajectory_tag: &str,
        tape_hash: &str,
        policy_hash: &str,
        constitution_hash: &str,
        code_hash: &str,
    ) -> Result<Self, V8CoreError> {
        let prov = TraceProvenance::new(tape_hash, policy_hash, constitution_hash, code_hash)?;
        Self::new(
            &episode.episode_id,
            trajectory_type,
            trajectory_tag,
            episode.as_of_time,
            prov,
        )
    }

    /// Convenience accessors for provenance fields
    pub fn tape_hash(&self) -> &str {
        &self.provenance.tape_hash
    }

    pub fn policy_hash(&self) -> &str {
        &self.provenance.policy_hash
    }

    pub fn constitution_hash(&self) -> &str {
        &self.provenance.constitution_hash
    }

    pub fn code_hash(&self) -> &str {
        &self.provenance.code_hash
    }

    /// Computes cryptographic hash for the entire trace context payload.
    pub fn compute_hash(&self) -> String {
        let mut c = Canon::new();
        c.push_str("EconomicTraceContext");
        c.push_str(self.trace_id.as_str());
        c.push_str(&self.opportunity_id);
        c.push_str(self.trajectory_type.as_str());
        c.push_str(&self.trajectory_tag);
        c.push_i64(self.pit_timestamp);
        c.push_str(&self.provenance.compute_hash());
        c.finish_blake3_hex()
    }
}

File: v8-next/src/v8_next/telemetry/identity.py
"""Canonical economic trace identity, trajectory modality and provenance — port of
``v8-core/src/telemetry/identity.rs`` (EEO-001H, D-136).

Constitutional invariants carried over from the Rust module:

1. **Opportunity sovereignty** — :class:`EconomicTraceId` identifies one execution or decision
   trajectory *of* an opportunity; it is not the opportunity identity itself.
2. **Trajectory identity** — the trajectory tag separates baseline, challenger and
   counterfactual branches while preserving opportunity identity alignment.
3. **Provenance isolation** — :class:`TraceProvenance` captures the tape/policy/constitution/code
   commit state without destroying identity alignment between those branches.
4. **Typed modality** — :class:`TrajectoryType` distinguishes ``Observed`` from
   ``Counterfactual`` without naming a heuristic.
5. **Identity lineage decoupling (EEO-002 invariant 5)** — :class:`EconomicTraceId`,
   :class:`SpanId`, ``BeliefReceiptId`` and :class:`TraceProvenance` are four *distinct* types
   over four distinct digest inputs. No identity in this module takes a wall clock: the only
   time any of them binds is the Point-In-Time decision timestamp the caller supplies, so the
   same content reproduces the same identity on any machine at any time.

Digest convention: the Rust side hashes a ``Canon`` byte encoding with BLAKE3. This port reuses
the repository's existing canonical-JSON helper (``v8_next.evaluation.store.canonical``) with
SHA-256. The bytes differ from the Rust digest, the property does not — same content gives the
same identity, different content gives a different one (``canonical`` rejects non-finite floats,
so a NaN fails closed instead of entering a digest as an anonymous number).
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Protocol

from v8_next.evaluation.store import canonical

#: The digest every identity in this package is computed with.
HASH_ALGORITHM = "sha256"


def canonical_digest(tag: str, payload: object) -> str:
    """Domain-separated digest of a canonical-JSON payload.

    ``tag`` names the identity domain (the Rust side pushes the same name as the first
    ``Canon::push_str``), so two identities with the same field values in different domains
    cannot collide. Wall-clock time is never part of ``payload`` at any call site in this
    package; a caller that needs a time dimension passes the Point-In-Time decision timestamp.
    """
    return hashlib.sha256(f"{tag}|{canonical(payload)}".encode()).hexdigest()


class TraceLineageError(Exception):
    """A trace identity, lineage or provenance contract was violated.

    The named failure every constructor in this package raises. ``code`` is the machine-readable
    discriminator (the Python counterpart of the Rust ``V8CoreError::TraceLineageError`` arm) and
    is carried on every subclass so a consumer branches on the code, never on message prose.
    """

    code: str = "TRACE_LINEAGE_ERROR"

    def __init__(self, message: str, *, code: str | None = None) -> None:
        resolved = code or type(self).code
        super().__init__(f"{resolved}: {message}")
        self.code = resolved


class TrajectoryType(StrEnum):
    """Execution modality for an economic decision trajectory."""

    #: Canonical observed forward simulation or physical execution.
    Observed = "Observed"
    #: Counterfactual branch under a registered or exploratory policy intervention.
    Counterfactual = "Counterfactual"

    def is_observed(self) -> bool:
        return self is TrajectoryType.Observed

    def is_counterfactual(self) -> bool:
        return self is TrajectoryType.Counterfactual

    def as_str(self) -> str:
        return str(self.value)


@dataclass(frozen=True)
class TraceProvenance:
    """Cryptographic provenance context binding the decision environment (D-136-RP-001 §6.1)."""

    tape_hash: str
    policy_hash: str
    constitution_hash: str
    code_hash: str

    @classmethod
    def new(
        cls,
        tape_hash: str,
        policy_hash: str,
        constitution_hash: str,
        code_hash: str,
    ) -> TraceProvenance:
        """Validate and construct a provenance bundle; an empty hash is refused.

        The Rust constructor returns ``Err(TraceLineageError)`` per empty field; this port raises
        the same named error with the same per-field message.
        """
        for field_name, value in (
            ("tape_hash", tape_hash),
            ("policy_hash", policy_hash),
            ("constitution_hash", constitution_hash),
            ("code_hash", code_hash),
        ):
            if not value:
                raise TraceLineageError(f"{field_name} cannot be empty in TraceProvenance")
        return cls(
            tape_hash=str(tape_hash),
            policy_hash=str(policy_hash),
            constitution_hash=str(constitution_hash),
            code_hash=str(code_hash),
        )

    def compute_hash(self) -> str:
        """Deterministic digest of the provenance bundle. Takes no time input."""
        return canonical_digest(
            "TraceProvenance",
            [self.tape_hash, self.policy_hash, self.constitution_hash, self.code_hash],
        )

    def as_dict(self) -> dict[str, Any]:
        return {
            "tape_hash": self.tape_hash,
            "policy_hash": self.policy_hash,
            "constitution_hash": self.constitution_hash,
            "code_hash": self.code_hash,
            "provenance_hash": self.compute_hash(),
        }


@dataclass(frozen=True, order=True)
class EconomicTraceId:
    """Canonical economic trace identifier (digest-derived hex).

    Identifies a specific execution/decision trajectory for an opportunity. Deliberately a
    distinct type from :class:`SpanId` and from ``BeliefReceiptId``: identity lineage decoupling
    (EEO-002 invariant 5) is a type-level property here, not a naming convention.
    """

    value: str

    @classmethod
    def new(cls, identifier: str) -> EconomicTraceId:
        return cls(str(identifier))

    @classmethod
    def compute(
        cls,
        opportunity_id: str,
        trajectory_tag: str,
        trajectory_type: TrajectoryType,
        pit_timestamp: int,
    ) -> EconomicTraceId:
        """Deterministic identity for an economic trajectory.

        The trajectory tag distinguishes baseline vs challenger vs counterfactual branches while
        preserving opportunity identity alignment. ``pit_timestamp`` is the Point-In-Time
        decision timestamp, never a wall clock.
        """
        return cls(
            canonical_digest(
                "EconomicTraceId",
                [opportunity_id, trajectory_tag, trajectory_type.as_str(), int(pit_timestamp)],
            )
        )

    def as_str(self) -> str:
        return self.value

    def __str__(self) -> str:
        return self.value


@dataclass(frozen=True, order=True)
class SpanId:
    """Canonical decision/evidence span identifier (digest-derived hex)."""

    value: str

    @classmethod
    def new(cls, identifier: str) -> SpanId:
        return cls(str(identifier))

    @classmethod
    def compute(
        cls,
        trace_id: EconomicTraceId,
        parent_span_id: SpanId | None,
        stage_name: str,
        start_time: int,
        disambiguator: str,
    ) -> SpanId:
        """Deterministic identity for a span.

        A span with no parent encodes an explicit ``null`` in the parent slot (the Rust
        ``Canon::push_null``), so "the root span of this stage at this time" and "the child of
        span X" cannot collide.
        """
        return cls(
            canonical_digest(
                "SpanId",
                [
                    trace_id.as_str(),
                    None if parent_span_id is None else parent_span_id.as_str(),
                    stage_name,
                    int(start_time),
                    disambiguator,
                ],
            )
        )

    def as_str(self) -> str:
        return self.value

    def __str__(self) -> str:
        return self.value


class OpportunityEpisodeLike(Protocol):
    """Structural view of the Rust ``OpportunityEpisode`` used by the ``from_episode`` helpers.

    The v8-next counterpart (``v8_next.opportunities.models.OpportunityRecord``) spells the two
    fields ``opportunity_id`` and ``as_of_time_ns``; the Rust struct spells them ``episode_id``
    and ``as_of_time``. The protocol names the Python spelling, and the mismatch is recorded here
    rather than papered over with a duck-typed getattr chain.
    """

    opportunity_id: str
    as_of_time_ns: int


@dataclass(frozen=True)
class EconomicTraceContext:
    """Immutable root context of an economic trace (D-136-RP-001 §6.1).

    Preserves opportunity identity, trajectory identity and provenance as three distinct
    dimensions. Construct through :meth:`new` (or the episode helpers) so the trace id is always
    derived from the content rather than accepted from a caller.
    """

    trace_id: EconomicTraceId
    opportunity_id: str
    trajectory_type: TrajectoryType
    trajectory_tag: str
    pit_timestamp: int
    provenance: TraceProvenance

    @classmethod
    def new(
        cls,
        opportunity_id: str,
        trajectory_type: TrajectoryType,
        trajectory_tag: str,
        pit_timestamp: int,
        provenance: TraceProvenance,
    ) -> EconomicTraceContext:
        """Validate and construct a context, deriving its trace identity from the content."""
        if not opportunity_id:
            raise TraceLineageError("opportunity_id cannot be empty")
        if not trajectory_tag:
            raise TraceLineageError("trajectory_tag cannot be empty")
        return cls(
            trace_id=EconomicTraceId.compute(
                opportunity_id, trajectory_tag, trajectory_type, pit_timestamp
            ),
            opportunity_id=str(opportunity_id),
            trajectory_type=trajectory_type,
            trajectory_tag=str(trajectory_tag),
            pit_timestamp=int(pit_timestamp),
            provenance=provenance,
        )

    @classmethod
    def from_episode(
        cls,
        episode: OpportunityEpisodeLike,
        tape_hash: str,
        policy_hash: str,
        constitution_hash: str,
        code_hash: str,
    ) -> EconomicTraceContext:
        """Context for an observed opportunity episode (Rust ``from_episode``)."""
        return cls.from_episode_trajectory(
            episode,
            TrajectoryType.Observed,
            "canonical_observed",
            tape_hash,
            policy_hash,
            constitution_hash,
            code_hash,
        )

    @classmethod
    def from_episode_trajectory(
        cls,
        episode: OpportunityEpisodeLike,
        trajectory_type: TrajectoryType,
        trajectory_tag: str,
        tape_hash: str,
        policy_hash: str,
        constitution_hash: str,
        code_hash: str,
    ) -> EconomicTraceContext:
        """Context for a named challenger or counterfactual trajectory (Rust helper)."""
        provenance = TraceProvenance.new(
            tape_hash, policy_hash, constitution_hash, code_hash
        )
        return cls.new(
            episode.opportunity_id,
            trajectory_type,
            trajectory_tag,
            episode.as_of_time_ns,
            provenance,
        )

    @property
    def tape_hash(self) -> str:
        return self.provenance.tape_hash

    @property
    def policy_hash(self) -> str:
        return self.provenance.policy_hash

    @property
    def constitution_hash(self) -> str:
        return self.provenance.constitution_hash

    @property
    def code_hash(self) -> str:
        return self.provenance.code_hash

    def compute_hash(self) -> str:
        """Digest of the whole trace context payload. Takes no wall-clock input."""
        return canonical_digest(
            "EconomicTraceContext",
            [
                self.trace_id.as_str(),
                self.opportunity_id,
                self.trajectory_type.as_str(),
                self.trajectory_tag,
                self.pit_timestamp,
                self.provenance.compute_hash(),
            ],
        )

    def as_dict(self) -> dict[str, Any]:
        return {
            "trace_id": self.trace_id.as_str(),
            "opportunity_id": self.opportunity_id,
            "trajectory_type": self.trajectory_type.as_str(),
            "trajectory_tag": self.trajectory_tag,
            "pit_timestamp": self.pit_timestamp,
            "is_observed": self.trajectory_type.is_observed(),
            "is_counterfactual": self.trajectory_type.is_counterfactual(),
            "provenance": self.provenance.as_dict(),
            "context_hash": self.compute_hash(),
        }
